scripts/xdist/jenkins_container_manager.py

```diff
@@ -0,0 +1,188 @@
+                    if retry == MAX_RUN_TASK_RETRIES:
+                        raise StandardError(
+                            "MAX_RUN_TASK_RETRIES ({}) reached while spinning up containers due to AWS throttling.".format(MAX_RUN_TASK_RETRIES)
+                        )
+                    logger.info("Hit error: {}. Retrying".format(err))
+                    countdown = 2 ** retry
+                    logger.info("Sleeping for {} seconds".format(countdown))
+                    time.sleep(countdown)
+                else:
+                    break
+
+            for task_response in response['tasks']:
+                task_arns.append(task_response['taskArn'])
+
+        not_running = task_arns
+        ip_addresses = []
+        all_running = False
+        for attempt in range(0, CONTAINER_RUN_TIME_OUT_MINUTES * 2):
+            time.sleep(30)
+            list_tasks_response = self.ecs.describe_tasks(cluster=self.cluster_name, tasks=not_running)['tasks']
+            del not_running[:]
+            for counter, task_response in enumerate(list_tasks_response):
+                if task_response['lastStatus'] == 'RUNNING':
+                    for container in task_response['containers']:
+                        ip_addresses.append(container["networkInterfaces"][0]["privateIpv4Address"])
+                else:
+                    not_running.append(task_response['taskArn'])
+
+            if not_running:
+                logger.info("Still waiting on {} containers to spin up".format(len(not_running)))
+            else:
+                logger.info("Finished spinning up containers")
+                all_running = True
+                break
+
+        if not all_running:
+            raise StandardError(
+                "Timed out waiting to spin up all containers."
+            )
+
+        logger.info("Successfully booted up {} containers.".format(number_of_containers))
+
+        ip_list_string = " ".join(ip_addresses)
+        logger.info("Container IP list: {}".format(ip_list_string))
+        ip_list_file = open("jenkins_container_ip_list.txt", "w")
+        ip_list_file.write(ip_list_string)
+        ip_list_file.close()
+
+        task_arn_list_string = " ".join(task_arns)
+        logger.info("Container task arn list: {}".format(task_arn_list_string))
+        task_arn_file = open("jenkins_container_task_arns.txt", "w")
+        task_arn_file.write(task_arn_list_string)
+        task_arn_file.close()
+
+    def terminate_containers(self, task_arns, reason):
+        """
+        Terminates jenkins-worker containers based on a list of task_arns.
+        """
+        for task_arn in task_arns:
+            response = self.ecs.stop_task(
+                cluster=self.cluster_name,
+                task=task_arn,
+                reason=reason
+            )
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(
+        description="JenkinsContainerManager, manages ECS containers in an AWS cluster."
```

Comment:
I sort of wonder if we should change all the "Jenkins" references in this file to "Pytest" instead; there isn't really anything Jenkins-specific about it, it should all work just as well from devstack or another CI system.  Probably not worth renaming images, etc. at this point though.

Reply:
That is a very good point. I'll do that to this file, I think keeping the AWS resources/docker image as jenkins-worker-container is OK, since that's where they will mostly be used. But I do agree that, in general, it's specific to pytest and not Jenkins.